dash/include/dash/halo/iterator/HaloStencilIterator.h

```diff
@@ -0,0 +1,517 @@
+#define DASH__HALO__ITERATOR__HALOSTENCILITERATOR_H
+
+#include <dash/Types.h>
+
+#include <dash/halo/Halo.h>
+
+#include <vector>
+
+namespace dash {
+
+enum class StencilViewScope : std::uint8_t { INNER, BOUNDARY, ALL };
+/*
+ * Iterator with stencil points and halo access \see HaloStencilOperator.
+ */
+template <typename ElementT, typename PatternT, typename StencilSpecT,
+          StencilViewScope Scope>
+class HaloStencilIterator {
+private:
+  static constexpr auto NumDimensions    = PatternT::ndim();
+  static constexpr auto NumStencilPoints = StencilSpecT::num_stencil_points();
+  static constexpr auto MemoryArrange    = PatternT::memory_order();
+  static constexpr auto FastestDimension =
+    MemoryArrange == ROW_MAJOR ? NumDimensions - 1 : 0;
+
+  using Self_t = HaloStencilIterator<ElementT, PatternT, StencilSpecT, Scope>;
+  using ViewSpec_t            = typename PatternT::viewspec_type;
+  using pattern_size_t        = typename PatternT::size_type;
+  using signed_pattern_size_t = typename std::make_signed<pattern_size_t>::type;
+  using RegionCoords_t        = RegionCoords<NumDimensions>;
+
+public:
+  // Iterator traits
+  using iterator_category = std::random_access_iterator_tag;
+  using value_type        = ElementT;
+  using difference_type   = typename PatternT::index_type;
+  using pointer           = ElementT*;
+  using reference         = ElementT&;
+
+  using HaloBlock_t     = HaloBlock<ElementT, PatternT>;
+  using HaloMemory_t    = HaloMemory<HaloBlock_t>;
+  using pattern_index_t = typename PatternT::index_type;
+  using region_index_t  = typename RegionCoords_t::region_index_t;
+  using LocalLayout_t =
+    CartesianIndexSpace<NumDimensions, MemoryArrange, pattern_index_t>;
+  using StencilP_t       = StencilPoint<NumDimensions>;
+  using ElementCoords_t  = std::array<pattern_index_t, NumDimensions>;
+  using StencilOffsets_t = std::array<signed_pattern_size_t, NumStencilPoints>;
+
+public:
+  /**
+   * Constructor
+   *
+   * \param haloblock \ref HaloBlock instance to use
+   * \param stencil_spec \ref StencilSpec to use
+   * \param stencil_offsets stencil offsets for every stencil point
+   * \param idx position of the iterator
+   */
+  HaloStencilIterator(const HaloBlock_t& haloblock, HaloMemory_t& halomemory,
+                      const StencilSpecT&     stencil_spec,
+                      const StencilOffsets_t& stencil_offsets,
+                      pattern_index_t         idx)
+  : _haloblock(haloblock), _halomemory(halomemory), _stencil_spec(stencil_spec),
+    _stencil_offsets(stencil_offsets),
+    _local_memory((ElementT*) _haloblock.globmem().lbegin()),
+    _local_layout(_haloblock.pattern().local_memory_layout()), _idx(idx) {
+    if(Scope == StencilViewScope::INNER)
+      set_view_local(_haloblock.view_inner());
+
+    if(Scope == StencilViewScope::ALL)
+      set_view_local(_haloblock.view_inner_with_boundaries());
+
+    if(Scope == StencilViewScope::BOUNDARY)
```

Comment:
Does C++ guarantee, that only valid values for `Scope` can be passed. If not, that we should have a `static_assert()` here to guarantee it.

Reply:
yes, it does (enum class)